File: backend/app/schemas/revise.py
from __future__ import annotations

from pydantic import BaseModel


class ReviseQuestion(BaseModel):
    ref: str
    kind: str  # "weak" | "refresher"
    module_id: str
    lesson_id: str
    concept: str
    question: str
    choices: list[str]


class ReviseSession(BaseModel):
    items: list[ReviseQuestion]


class ReviseModule(BaseModel):
    module_id: str
    title: str
    icon: str
    topic: str
    due_weak_count: int


class ReviseAnswerIn(BaseModel):
    ref: str
    selected_index: int


class ReviseAnswerResult(BaseModel):
    correct: bool
    answer_index: int
    explanation: str
    xp_awarded: int
    goal_met: bool

File: frontend/src/api/revise.ts
import { useQuery } from '@tanstack/react-query';
import { apiFetch } from './client';

export type ReviseQuestion = {
  ref: string;
  kind: 'weak' | 'refresher';
  module_id: string;
  lesson_id: string;
  concept: string;
  question: string;
  choices: string[];
};
export type ReviseSession = { items: ReviseQuestion[] };

export type ReviseModule = {
  module_id: string;
  title: string;
  icon: string;
  topic: string;
  due_weak_count: number;
};

export type ReviseAnswerResult = {
  correct: boolean;
  answer_index: number;
  explanation: string;
  xp_awarded: number;
  goal_met: boolean;
};

export const reviseApi = {
  getModules: () => apiFetch<ReviseModule[]>('/revise/modules'),
  getSession: (moduleId?: string) =>
    apiFetch<ReviseSession>(
      moduleId ? `/revise/session?module_id=${moduleId}` : '/revise/session',
    ),
  postAnswer: (ref: string, selectedIndex: number) =>
    apiFetch<ReviseAnswerResult>('/revise/answer', {
      method: 'POST',
      body: JSON.stringify({ ref, selected_index: selectedIndex }),
    }),
};

export function useRevisableModules() {
  return useQuery({
    queryKey: ['revise-modules'],
    queryFn: () => reviseApi.getModules(),
    retry: false,
    staleTime: 60_000,
  });
}
